backend/webdav/webdav.go

```diff
@@ -1154,19 +1163,146 @@ func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, op
 		return errors.Wrap(err, "Update mkParentDir failed")
 	}
 
-	size := src.Size()
+	if o.fs.canChunk {
+		err = o._UpdateChunked(ctx, in, src, options...)
+		if err != nil {
+			return err
+		}
+	} else {
+		size := src.Size()
+		contentType := fs.MimeType(ctx, src)
+		filePath := o.filePath()
+		extraHeaders := map[string]string{}
+		if o.fs.useOCMtime || o.fs.hasMD5 || o.fs.hasSHA1 {
+			if o.fs.useOCMtime {
+				extraHeaders["X-OC-Mtime"] = fmt.Sprintf("%d", src.ModTime(ctx).Unix())
+			}
+			// Set one upload checksum
+			// Owncloud uses one checksum only to check the upload and stores its own SHA1 and MD5
+			// Nextcloud stores the checksum you supply (SHA1 or MD5) but only stores one
+			if o.fs.hasSHA1 {
+				if sha1, _ := src.Hash(ctx, hash.SHA1); sha1 != "" {
+					extraHeaders["OC-Checksum"] = "SHA1:" + sha1
+				}
+			}
+			if o.fs.hasMD5 && extraHeaders["OC-Checksum"] == "" {
+				if md5, _ := src.Hash(ctx, hash.MD5); md5 != "" {
+					extraHeaders["OC-Checksum"] = "MD5:" + md5
+				}
+			}
+		}
+		err = o._UpdateSimple(ctx, in, filePath, size, contentType, extraHeaders,  options...)
+		if err != nil {
+			return err
+		}
+	}
+
+	// read metadata from remote
+	o.hasMetaData = false
+	return o.readMetaData(ctx)
+}
+
+
+// Standard update 
+func (o *Object) _UpdateSimple(ctx context.Context, in io.Reader, filePath string, size int64, contentType string, extraHeaders map[string]string, options ...fs.OpenOption) (err error) {
+
 	var resp *http.Response
 	opts := rest.Opts{
 		Method:        "PUT",
-		Path:          o.filePath(),
+		Path:          filePath,
 		Body:          in,
 		NoResponse:    true,
 		ContentLength: &size, // FIXME this isn't necessary with owncloud - See https://github.com/nextcloud/nextcloud-snap/issues/365
-		ContentType:   fs.MimeType(ctx, src),
+		ContentType:   contentType,
+		Options:       options,
+		ExtraHeaders:  extraHeaders,
+	}
+
+	err = o.fs.pacer.CallNoRetry(func() (bool, error) {
+		resp, err = o.fs.srv.Call(ctx, &opts)
+		return o.fs.shouldRetry(resp, err)
+	})
+	if err != nil {
+		// Give the WebDAV server a chance to get its internal state in order after the
+		// error.  The error may have been local in which case we closed the connection.
+		// The server may still be dealing with it for a moment. A sleep isn't ideal but I
+		// haven't been able to think of a better method to find out if the server has
+		// finished - ncw
+		time.Sleep(1 * time.Second)
+		// Remove failed upload
+		_ = o.Remove(ctx)
+		return err
+	}
+	return nil
+
+}
+
+
+// Chunked update for Nextcloud (see https://docs.nextcloud.com/server/20/developer_manual/client_apis/WebDAV/chunking.html)
+func (o *Object) _UpdateChunked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+
+	var uploadDir string = strings.ReplaceAll(path.Join(o.fs.root,o.remote), "/", "_")
```

Comment:
This could potentially be very long - is that OK? As in limit for most file systems is 255 chars per file element, so this could easily be > 255 chars long.

Using the md5sum of the joined path might be better with a prefix `chunked-upload-` as these will be visible to the user if only temporarily I think

Reply:
✅